Major refactor

sdfgeoff opened the pull request:
So this is a major refactoring designed to make the system easier to work on in the future. The most significant changes are:

 - There is now the concept of a "converter" which handles a single conversion. These are in the converters folder. They convert from a blender concept into a godot concept. Often this is a blender object into a node, but it can also be into a resource. Converters call other converters as necessary (eg a mesh object converter calls the mesh resource converter which calls the material converter)
 - The conversion is slightly separated from the actual file format. The files "structures.py" and "encoders.py" handle the final stage of converting from python objects into ascii. Everything is converted ad-hoc, and then at the very end, everything is serialized and written to the file. One of the structures is the "ESCNFile" which keeps track of resource ID's
 - Removal of the commented out code. There was lots of commented out code (that looks like it targeted an older xml file format). This has been removed.

Features Added:
 - Export of physics objects (except cone and cylinder, which Godot does not support).
 - Optionally search through the export or project directory for godot materials of matching names. This frequently prevents needing inherited scenes of your exported scenes. (in the limited use so far, this has become extremely extremely useful).
 - Primitive material export.

Other:
 - Creation of a primitive test-system. Run "make test-blends" and it will export all the blend files in a test project. This allows visual comparison from blender -> godot, and will hopefully make finding/fixing issues easier.
 - It is now mostly pep8 and pylint compliant. Still a few non-compliant variable names, but it scores a 9.5/10, which I am pretty happy with.
 - Fixed about a dozen issues with lights and cameras.

Things to do before merge:
- [x] See if I can merge the Resources and Node classes together, they all do pretty much the same.
- [x] See if the manual-writing into the mesh SubResource can be replaced by something more structured.


If anyone has time, I'd appreciate a quick code review. In particular about if the overall code structure appears sane.

rminderhoud commented on io_scene_godot/__init__.py:
```diff
@@ -71,92 +79,77 @@ class ExportGodot(bpy.types.Operator, ExportHelper):
 
 
 def menu_func(self, context):
+    """Add to the manu"""
```
Typo manu -> menu

rminderhoud commented on io_scene_godot/converters/mesh.py:
```diff
@@ -0,0 +1,339 @@
+        return surface_lines
+
+
+CMP_EPSILON = 0.0001
```
Why is this constant randomly here?

sdfgeoff commented:
Thanks for the review. I've fixed those two points.

Automated testing is, hmm, interesting. We can detect if the software throws an error, but to test if the export actually worked is nearly impossible.  In the most recent commit I have added an entry to the Makefile `test-import` which will build the blends, and will open godot to let it import them. It then scans godot's output for the string `ERROR`. This perhaps isn't the most robust solution, but it is all I could think of for getting some sort of pass/fail.

The next challenge is setting up something to automate this. With gitlab, you can set up a gitlab runner, but I haven't found anything for github. We will need a custom machine to run this on as we require both blender and Godot.
I do have a machine here that could be used as a build runner, but I am not sure what the best approach is to integration with gitlab. Any advice?

rminderhoud commented:
Refactor of `InternalResource`/`ExternalResource`/`NodeTemplate` looks good. 

Yeah the dependency on blender is tricky. Ideally I was thinking something like: Import blend into blender -> Export ESCN -> Diff generated ESCN vs. known output. My assumption is that the plugin output is deterministic for a given blend file. What do you think?

Since the project is open source we can use https://docs.travis-ci.com/. We'd need to figure out how to bootstrap the runner to include blender and the addon.

rminderhoud commented:
@sdfgeoff you are a beast

sdfgeoff commented:
I think that method of testing is quite sane. I've implemented it, and you can now run `make compare` which will generate a diff. Or you can just run `make` which does the comparison and does a style test (which now almost passes).

I set up a travis build, only to discover that apparently it is not quite deterministic. I fixed what I could (argument order for nodes), but I can't figure out why blender is iterating through vertices in one order on one PC, and another order on the build server. I don't think there's much I can do about that. However, it is only for one single blend file, so one option is to remove it and hope no other blends exhibit that problem.

Anyway, I think I'm done with my refactoring, as I'm now happy with the structure. From here I think it's mostly a case of adding more exporter features. So ... _I think this branch is ready to merge_.

sdfgeoff commented:
Added some documentation on the fork's wiki:
https://github.com/simedis/godot-blender-exporter/wiki

rminderhoud commented:
I feel that this should be merged if @sdfgeoff is ready:

- Refactored code base
- Added functionality
- Added tests + CI
- New documentation

 @akien-mga

sdfgeoff commented:
Yup, I think this should be merged, when @akien-mga has some free time....

akien-mga commented:
Well I wanted @reduz to review it, but since we have so much backlog, I'll trust you guys on that one.
